File: my-solutions/0019-remove-nth-node-from-end-of-list/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: Optional[ListNode]
        :type n: int
        :rtype: Optional[ListNode]
        """
        fast = slow = head
        prev = None
        i = 0
        while fast:
            fast = fast.next
            if i >= n:
                prev = slow
                slow = slow.next

            i += 1

        if prev == None:
            return head.next

        prev.next = slow.next

        return head        

File: my-solutions/0019-remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* last = head;
        ListNode* first = nullptr;
        int i = 0;

        while(last) {
            last = last->next;
            if(first) first = first->next;
            ++i;
            if(i == n + 1) {
                first = head;
            }
        }

        if(i == n) {
            ListNode* temp = head;
            ListNode* newHead = head->next;
            temp->next = nullptr;
            delete(temp);
            return newHead;
        }

        if(!first) {
            return head;
        } else if (!first->next) {
            return head;
        } else if(!first->next->next) {
            first->next = nullptr;
            return head;
        } else {
            ListNode* temp = first->next;
            first->next = temp->next;
            temp->next = nullptr;
            delete(temp);
            return head;
        }
    }
};
